Reviewed the custom authenticator docs (#2884)
wso2/docs-is: en/docs/develop/writing-a-custom-local-authenticator.md

# Writing a Custom Local Authenticator

> The default WSO2 Identity Server authenticator is the basic authenticator. It is a local authenticator that authenticates the end users using a connected user store and the provided user name and password. When a user needs to be authenticated with the user store and authorized based on the assigned user role, you can write a custom local authenticator. Let's try out a sample scenario on this.

Reviewer: ```
The default WSO2 Identity Server authenticator is the basic authenticator. It is a local authenticator that authenticates the end users using a connected user store and the provided user name and password.
```
this more meaningful

> A basic authenticator is available in the WSO2 Identity Server by default. A basic authenticator authenticates end users using a connected userstore and the provided username and password.

Reviewer 2: ```suggestion
A basic authenticator is available in the WSO2 Identity Server by default. A basic authenticator authenticates end users using a connected user store and the provided username and password.
```

With the WSO2 Identity Server, you can write your own local authenticator to define various authentication logic.

### Sample scenario

> You have the **playground2** app that you want users to log in with their telephone numbers instead of usernames. Once a user enters a telephone number, your authentication logic should identifiy the user and validate the user's credentials. 

Reviewer 2: Is it the **playground2** app or the **pickup-dispatch** app?
@himeshsiriwardana

                    }
                    throw new InvalidCredentialsException("User authentication failed due to invalid credentials",
                            User.getUserFromUserName(username));
                }

                // When the user is successfully authenticated, add the user to the authentication context to be used later in
                // the process.
                if (user != null) {
                    username = user.get().getUsername();
                    authenticationContext.setSubject(AuthenticatedUser.createLocalAuthenticatedUserFromSubjectIdentifier(username));
                }
            }

            @Override
            public String getContextIdentifier(HttpServletRequest httpServletRequest) {

                return httpServletRequest.getParameter("sessionDataKey");
            }

            @Override
            public String getName() {

                return "SampleLocalAuthenticator";
            }

            @Override
            public String getFriendlyName() {

                return "sample-local-authenticator";
            }
        }
        ```    

>         !!! note    

Reviewer: ```suggestion
        !!! Note    
```

                if (log.isDebugEnabled()) {
                    log.debug("SampleLocalAuthenticator bundle is deactivated");
                }
            }

            public static RealmService getRealmService() {

                return realmService;
            }

            @Reference(name = "realm.service",
                    service = org.wso2.carbon.user.core.service.RealmService.class,
                    cardinality = ReferenceCardinality.MANDATORY,
                    policy = ReferencePolicy.DYNAMIC,
                    unbind = "unsetRealmService")
            protected void setRealmService(RealmService realmService) {

                if (log.isDebugEnabled()) {
                    log.debug("Setting the Realm Service");
                }
                SampleLocalAuthenticatorServiceComponent.realmService = realmService;
            }

            protected void unsetRealmService(RealmService realmService) {

                if (log.isDebugEnabled()) {
                    log.debug("UnSetting the Realm Service");
                }
                SampleLocalAuthenticatorServiceComponent.realmService = null;
            }
        }
        ```

> 4.  Build the project using maven. (Click to see the sample project

Reviewer 2: ```
4. Build the project using maven. 

      !!! Info
          See the [sample project](https://github.com/wso2/samples-is/tree/master/authenticators/components/org.wso2.carbon.identity.sample.local.authenticator)) for details.
```

    [here](https://github.com/wso2/samples-is/tree/master/authenticators/components/org.wso2.carbon.identity.sample.local.authenticator)).

> 5.  Copy and paste the .jar file

Reviewer 2: ```
5. Copy the .jar file (`org.wso2.custom.authenticator.local-1.0.0.jar `) from the **target** folder in your project directory and copy it to the `<IS_HOME>/repository/components/dropins ` folder.
```

Is this .jar file name correct?

Reviewer: `org.wso2.carbon.auth.local.sample-1.0.0.jar`

    `           org.wso2.custom.authenticator.local-1.0.0.jar          `
    inside 
    `           <IS_HOME>/repository/components/dropins          `
    folder.

6.  Start WSO2 Identity Server.

7.  Sign in to the **Management Console**.

> 8.  Create a new user named `Larry`.

Reviewer: do we need to link docs on how to create the user and update the profile?

Reviewer 2: yes +1

> 9.  Go to the user's profile and update the telephone number.

Reviewer: Mention that it's Lary's profile

                    policy = ReferencePolicy.DYNAMIC,
                    unbind = "unsetRealmService")
            protected void setRealmService(RealmService realmService) {

                if (log.isDebugEnabled()) {
                    log.debug("Setting the Realm Service");
                }
                SampleLocalAuthenticatorServiceComponent.realmService = realmService;
            }

            protected void unsetRealmService(RealmService realmService) {

                if (log.isDebugEnabled()) {
                    log.debug("UnSetting the Realm Service");
                }
                SampleLocalAuthenticatorServiceComponent.realmService = null;
            }
        }
        ```

4.  Build the project using maven. See the [sample project](https://github.com/wso2/samples-is/tree/master/authenticators/components/org.wso2.carbon.identity.sample.local.authenticator) for details.

5.  Copy and paste the .jar file
    `           org.wso2.carbon.auth.local.sample-1.0.0.jar          `
    inside 
    `           <IS_HOME>/repository/components/dropins          `
    folder.

6.  Start WSO2 Identity Server.

7.  Sign in to the **Management Console**.

> 8.  Create a new user named `Larry`. Learn how to [Create a user](https://is.docs.wso2.com/en/latest/learn/adding-users-and-roles/#create-a-user)

Reviewer 2: link has to be the relative link.

9.  Go to `Larry`'s profile and update the telephone number.

10. Create a service provider for the application.  
    (You must have the app deployed in the tomcat server.)

11. In the service provider configuration, under **Inbound Authentication Configuration** section, click **OAuth/OpenID Connect Configuration > Configure**. 
    ![OAuth/OpenID Connect Configuration option](../assets/img/using-wso2-identity-server/oauth-openid-connect-configuration-option.png) 

12. Provide the `callback URL` and register it as an OAuth2 client app.

    ![Callback url](../assets/img/using-wso2-identity-server/app-callback-url.png) 

13. Under **Local & Outbound Authentication Configuration**, select **Local
    authentication**.
    
    ![Local authentication check box](../assets/img/using-wso2-identity-server/local-authentication-check-box.png) 

    !!! note
        On the corresponding drop-down list, you can see, `sample-local-authenticator` which is the display name of the custom authenticator that was written. Thus, you can make sure your custom authenticator is there and ready for use.    

> 14. Configure the pickup-dispatch app with the appropriate **Client ID**. Learn how to [deploy the pickup-dispatch webapp](https://is.docs.wso2.com/en/5.11.0/learn/deploying-the-sample-app/#deploying-the-pickup-dispatch-webapp) 

Reviewer 2: break this into two.
CLient Id and Client secret

13. Under **Local & Outbound Authentication Configuration**, select **Local

>     authentication** check box.

Reviewer 2: ```
13. Under **Local & Outbound Authentication Configuration**, select **Local
    authentication**.
```
It should be the above

    ![Local authentication check box](../assets/img/using-wso2-identity-server/local-authentication-check-box.png) 

    !!! note

>         On the corresponding drop-down list, you can see, `sample-local-authenticator` which is the display name of the custom authenticator that was written. Thus, you can make sure your custom authenticator is there and ready for use.    

Reviewer 2: ```suggestion
        On the corresponding drop-down list, you can see `sample-local-authenticator`, which is the display name of the custom authenticator that was written. Thus, you can make sure your custom authenticator is there and ready for use.    
```

Reviewer: "custom authenticator is deployed correctly and ready for use. "

> 14. Visit the playground2 app and provide the **Client ID** of the registered

Reviewer 2: playground2 app?

    app and give the **Scope** as `openid` to make sure it is
    using the OpenID Connect flow. Click **Authorize**.  
    ![Playground app](../assets/img/using-wso2-identity-server/playground-app.jpeg)
    Note that you will be directed to WSO2 Identity Server login page.

> 15. Provide the telephone number as the user identifier and enter the password of the user `Larry`. You are prompted to approve the app and log in.

Reviewer: for the first time, the consent page also can be there
